### ISSUE TITLE
24 bit FLAC files decoded with FFMPEG are silent

### ISSUE THREAD

**Issue by fladd** [pyglet/pyglet#404]
When playing 24 bit FLAC files as a FFmpegSource, they do play, but there is no sound.

**Comment by caffeinepills** [comment]
I want to say it was created to only support up to 16 bit audio. 

https://github.com/pyglet/pyglet/blob/916b9563fd68dcefa89d225b02b79a7b838496ec/pyglet/media/codecs/ffmpeg.py#L513

You can try modifying this line to `(8, 16, 32)` instead of just `(8, 16)`. It should play, but would have to be investigated if this is supported on all OS's and supported ffmpeg versions without issues.

**Comment by fladd** [comment]
Thanks, I tried that change, but it did not work (I also tried to add `24`). Is there a reason to not support 24 bit? As far as I know, FFmpeg does support it.

**Comment by caffeinepills** [comment]
I just tested this and was able to play a 24 bit FLAC's (from test FLACs) with that line change on Windows. Can you post an example of a FLAC having issues?

**Comment by fladd** [comment]
Ah, you are right. I tested on MacOS, which only has the OpenAL driver, which apparently cannot do more than 16 bit. Hence, on Windows (DirectSound) and Linux (PulseAudio), this should work.

That's an easy fix. For now I can do this locally, but I hope it gets fixed on the Pyglet side soon.

**Comment by benmoran56** [comment]
I looked into this a bit, and I think I understand why 24-bit support wasn't supported. 
According to what I've read,  FFmpeg uses a 32bit type when decoding 24-bit audio. 
What appears to happen, is the data is simply shifted to the left 8 bits, padding up to 32bits/4bytes. 
It _might_ be as easy as dropping every Nth byte, but it needs investigating.

**Comment by benmoran56** [comment]
Yeah, I did a quick hack and I can confirm that this is indeed the case. With a quick hack, I was able to play a 24bit FLAC file successfully.  I'll see if I can make a more permanent fix.

**Comment by fladd** [comment]
> Yeah, I did a quick hack and I can confirm that this is indeed the case. With a quick hack, I was able to play a 24bit FLAC file successfully. I'll see if I can make a more permanent fix.

You mean with the OpenAL driver? Could you let me know how you made that work, then I could implement this already in my application by subclassing the relevant class in Pyglet, before it gets officially fixed.

Thanks.
